ssl/statem/extensions.c

```diff
@@ -929,9 +929,27 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
         ret = s->session_ctx->ext.servername_cb(s, &altmp,
                                        s->session_ctx->ext.servername_arg);
 
-    if (!sent) {
-        OPENSSL_free(s->session->ext.hostname);
-        s->session->ext.hostname = NULL;
+    /*
+     * For servers, propagate the SNI hostname from the temporary
+     * storage in the SSL to the persistent SSL_SESSION, now that we
+     * know we accepted it.
+     * Clients make this copy when parsing the server's response to
+     * the extension, which is when they find out that the negotiation
+     * was successful.
+     */
+    if (s->server) {
```

Comment:
When final_server_name() is called at the client side,
Should we consider coping the servername from s->ext.hostname to s->session->ext.hostname  ?
We may consider doing it, in case 'sent' is true. (At the client side, 'sent' is true when the client gets the "sni ack" from the server).